### PR TITLE
fix(cli): convert Palmer inputs to the inches palmer.pdsi() takes (#960)

### LINKED ISSUE #960: CLI Palmer path feeds mm precipitation/PET and unconverted AWC to palmer.pdsi()
**Describe the bug**

`--index palmers` computes with mixed units. `palmer.pdsi()` documents precipitation, PET, and AWC in **inches** (`src/climate_indices/palmer.py`), but the CLI normalizes precipitation and PET to **millimeters** and never touches AWC:

- `src/climate_indices/__main__.py` (`_compute_write_index`, "precipitation and PET should be in millimeters"): inch-labeled precip/PET variables are multiplied by 25.4; mm-labeled variables pass through unchanged.
- AWC is copied from its netCDF variable with no unit handling or validation (`_compute_write_index` palmers branch; `_validate_args` checks AWC dimensions only).
- `_palmers()` then hands those arrays to `palmer.pdsi()`, which treats them as inches while the internal water-balance constants (`AWCTOP = 1.0`) are inches.

As a result every CLI Palmer run feeds either mm precipitation/PET with inch AWC (inch-labeled input) or values-in-wrong-units with inch AWC (mm-labeled input). There is no input labeling that produces a scientifically consistent Palmer computation through the CLI.

**To Reproduce**

1. Write a divisions netCDF with precipitation and PET in inches (`units="inches"`), as nClimDiv-style inputs ship, and AWC in inches.
2. Run `climate_indices --index palmers --periodicity monthly --calibration_start_year 1981 --calibration_end_year 2010 --netcdf_precip precip.nc --var_name_precip precip --netcdf_pet pet.nc --var_name_pet pet --netcdf_awc awc.nc --var_name_awc awc --output_file_base out`.
3. Compare `out_pdsi.nc` against `palmer.pdsi(inch_precips, inch_pet, awc, ...)`: they differ. The CLI output instead matches `palmer.pdsi(inch_precips * 25.4, inch_pet * 25.4, awc, ...)`, i.e. the mm/inches mismatch.

Alternatively, label the same data `units="mm"`: the CLI skips conversion and passes mm values to `palmer.pdsi()` as if they were inches.

**Expected behavior**

One documented unit convention for Palmer CLI inputs. Either:

- **A — inches end to end**: declare CLI Palmer precipitation, PET, and AWC as inches (matching `palmer.pdsi()`), skip the inches→mm conversion for `--index palmers`/`all`, validate the AWC variable's units, and update the examples/docs. (Likely the smaller change.)
- **B — millimeters at the CLI boundary**: keep the CLI's mm convention, convert precipitation and PET from mm to inches before calling `palmer.pdsi()`, and convert or require AWC in inches.

Either way the CLI output must match `palmer.pdsi()` called with the same physical values in the documented unit.

**Additional context**

- Surfaced while adding the CLI end-to-end suite (issue #918, PR #956). That suite writes **mm**-labeled inputs and compares the CLI output against `palmer.pdsi()` on the same mm arrays, so it covers dispatch/reshape only and cannot detect this unit mismatch; it will need updating with the fix.
- Existing Palmer CLI tests (`tests/test_main_palmers.py`) feed the inch-based fixture arrays directly to `_palmers()`/`_parallel_process()`, bypassing `_compute_write_index()`'s conversion, so they also do not cover it.
- When fixed, keep the validated reference/oracle assertions and tolerances untouched (per the test-suite cleanup rules in #908).

**Acceptance criteria**

- [ ] A single unit convention for CLI Palmer inputs (precipitation, PET, AWC) is chosen and documented.
- [ ] CLI Palmer output matches `palmer.pdsi()` on the same physical inputs in that convention.
- [ ] Inch-labeled and mm-labeled inputs each reach `palmer.pdsi()` in consistent units, or unsupported labels are rejected with a clear message.
- [ ] An end-to-end test exercises the conversion/validation branch (nClimDiv-style inches).

### PR DESCRIPTION
**Pull request by monocongo**
Closes #960.

## What

`--index palmers` fed `palmer.pdsi()` mixed units: precipitation and PET were normalized to millimeters (as for every other index) and available water capacity was passed through untouched, so no input labeling produced a scientifically consistent Palmer computation.

The CLI keeps its millimeter convention and converts the three Palmer inputs to the inches `palmer.pdsi()` takes:

- precipitation and PET are normalized to mm as before, then converted to inches in the Palmer path;
- AWC labeled `mm` / `millimeters` / `millimeter` is converted to inches;
- AWC with no `units` attribute is assumed to already be inches, preserving the previous interpretation of unitless AWC files and matching the fire adapters' unit handling;
- any other AWC unit raises `ValueError: Unsupported available water capacity units: ...`.

**Decision — option B (millimeters at the CLI boundary), not option A (inches end to end).** The CLI documents and tests millimeters for precipitation and PET across every index, and the documented Palmers example uses nClimGrid metric inputs; making one index demand inches would split the CLI's input contract. Option A remains a small change from here if you prefer it.

## Tests

`tests/test_cli_e2e.py`:

- the existing end-to-end Palmers test now runs all four combinations of mm/inch-labeled precipitation and PET and AWC, and asserts the CLI output equals `palmer.pdsi()` on the same physical inputs in inches;
- a new test asserts an AWC variable with an unsupported unit is rejected.

Full gate: `uv run ruff check src/ tests/`, `uv run ruff format --check src/ tests/`, `uv run mypy src/`, `uv run pytest` — 2060 passed.

## Docs

- `docs/index.rst` Palmers section documents the accepted input units;
- `docs/component-inventory.md` unit-conversion section updated, `llms-full.txt` regenerated.

## Acceptance

- [x] one unit convention for CLI Palmer inputs chosen and documented
- [x] CLI Palmer output matches `palmer.pdsi()` on the same physical inputs in that convention
- [x] inch-labeled and mm-labeled inputs each reach `palmer.pdsi()` in consistent units; unsupported labels rejected with a clear message
- [x] end-to-end test exercises the conversion and validation branches

## Summary by Sourcery

Ensure CLI Palmer calculations use consistent physical units and clearly validate unsupported input labels.

Bug Fixes:
- Convert Palmer precipitation, PET, and available water capacity inputs to the inches required by the Palmer calculations while preserving the CLI’s millimeter convention.
- Reject unsupported available water capacity units and precipitation rate labels for Palmer computations.

Documentation:
- Document Palmer input unit handling and conversion behavior.

Tests:
- Expand Palmer end-to-end coverage across millimeter- and inch-labeled inputs and add validation tests for unsupported units and precipitation rates.

## Summary by CodeRabbit

* **Bug Fixes**
  * Palmer index calculations now correctly handle precipitation, potential evapotranspiration, and available water capacity provided in inches or millimeters.
  * Available water capacity without a units attribute is treated as inches.
  * Unsupported available water capacity units are rejected without producing output.
  * Precipitation labeled as a daily rate (`mm/dy`) is rejected because Palmer indices require monthly accumulated precipitation.

* **Documentation**
  * Added guidance describing supported units, conversion behavior, and precipitation requirements for Palmer index inputs.

### CODE REVIEW

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/973 (`fix/960-palmer-cli-units` @ 1f997581 against `main`)
- **Lanes run:** 3 read-only lanes — dataflow/units correctness, test adequacy, maintainability/contracts (all on DeepSeek models)
- **Validation:** `uv run ruff check src/ tests/` pass; `uv run ruff format --check src/ tests/` pass; `uv run mypy src/` pass; `uv run pytest` 2062 passed; `uv run pytest tests/test_cli_e2e.py -k palmers -q` 7 passed; pre-fix mutation check by the tests lane ran HEAD's palmers tests against `origin/main` sources → 5 failed (every parametrized combo, including inch-labeled inputs, and the rejection case), confirming the new tests detect the original bug.
- **Fix commits:** [1f997581 — fix(cli): harden the Palmer unit conversion for integer inputs](https://github.com/monocongo/climate_indices/commit/1f997581d351db2ede3f3676c7df0f2b6ffe3e4a)

### Findings and fix status

#### [P3] Missing coverage of the absent-`units` AWC path — Fixed
- **Location:** `src/climate_indices/__main__.py` (AWC unit handling), `tests/test_cli_e2e.py` (parametrization)
- **Origin:** test adequacy; maintainability (same claim found independently by both lanes)
- **Problem:** the rewritten end-to-end test always wrote a `units` attribute on the AWC variable, so the documented "absent `units` ⇒ assumed inches" contract — previously covered by the old test writing a bare AWC dataset — had no coverage at all.
- **Applied fix:** `awc_units` now parametrizes over `mm`, `inches`, and `None` (no attribute); commit `1f997581`.

#### [P4] Integer-valued variables crashed the new conversion — Fixed
- **Location:** `src/climate_indices/__main__.py`, Palmer precip/PET and AWC conversion
- **Origin:** dataflow/units
- **Problem:** in-place `/= 25.4` on an `int64` variable raises `UFuncTypeError: Cannot cast ufunc 'divide' output from dtype('float64') to dtype('int64')`. Previously reachable only for inch-labeled integer input; the new mm→inch division made it reachable for all-mm Palmer inputs and for mm-labeled AWC (verified with a direct probe).
- **Applied fix:** divide out-of-place so the variable is promoted to float; commit `1f997581`.

#### [P4] Palmer docs did not state the absent-`units` AWC rule — Fixed
- **Location:** `docs/index.rst` (Palmers monthly note)
- **Origin:** maintainability
- **Problem:** the note read as if AWC only had to be inches; it omitted both the accepted mm conversion and the absent-attribute rule that the code implements.
- **Applied fix:** note now states that AWC without a `units` attribute is assumed to be inches and that any other declared unit is rejected; commit `1f997581`.

#### [P3] `--index all` Palmer unit conversion is not exercised end to end — Blocked (deferred)
- **Location:** `tests/test_cli_e2e.py`
- **Origin:** test adequacy
- **Problem:** the conversion keys off `keyword_arguments["index"]`, which `process_climate_indices` sets to `"palmers"` for the `all` pass, so the covered `--index palmers` test runs the same branch — but nothing proves it, and a future refactor keying on the CLI-level `arguments.index` would silently restore the bug for `--index all` users.
- **Blocker:** covering it end to end needs an `--index all` run, which additionally requires a temperature input and computes SPI/SPEI/PNP/scaled/PET, turning one focused test into a full-workflow test. Not warranted inside this fix; follow-up is to extend `tests/test_cli_e2e.py` with an `all`-driven case when that workflow fixture exists.

#### [P4] Unsupported-AWC-units error fires after the shared-memory copy — Blocked (declined)
- **Location:** `src/climate_indices/__main__.py`, `_compute_write_index` Palmer branch
- **Origin:** maintainability
- **Problem:** the `ValueError` is raised after precip/PET have been copied into shared memory, so a rejected AWC file wastes one full-array copy before failing.
- **Blocker:** the only validation home that fails earlier is `_validate_args()`, which does not perform the conversion; moving the check there splits the accepted-unit policy across two sites, where adding an mm spelling in one and not the other would silently feed wrong values — a worse failure than a wasted copy on an error path that writes no output and logs a clear message.

### Reviewed and dismissed (no action)
- Inches→mm→inches round trip is not bit-exact (~1 ulp on ~15% of samples, max 8.9e-16) — no measurable effect on Palmer outputs; the alternative restructures the shared unit normalization for a P4.
- `units="kg m-2"` AWC is now rejected rather than silently passed through — deliberately stricter per issue #960's acceptance criteria.
- The PET file label is driven by the same parametrization value as precipitation — correct today given both describe one CLI convention.

### Remaining blockers
None — no unresolved P0/P1; the two Blocked items are P3/P4 and do not block merge.

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/973
- **Lanes run:** 2 (correctness/edge-case; test-quality/coverage)
- **Validation:** `uv run ruff check src/ tests/` (clean), `uv run ruff format --check src/ tests/` (clean), `uv run mypy src/` (no issues), `uv run pytest` (2073 passed)
- **Fix commits:** [`1f99758`](https://github.com/monocongo/climate_indices/commit/1f997581d351db2ede3f3676c7df0f2b6ffe3e4a) fix(cli): harden the Palmer unit conversion for integer inputs, [`cc590e5`](https://github.com/monocongo/climate_indices/commit/cc590e52) test(cli): strengthen Palmer unit-conversion test coverage

### Findings and fix status

#### [P2] In-place division crashes on integer-dtype precip/PET/AWC input — Fixed
- **Location:** `src/climate_indices/__main__.py` (Palmer precip/PET conversion, and the AWC conversion further down)
- **Origin:** Correctness/edge-case lane
- **Problem:** The conversion used in-place true division (`dataset[var].values /= 25.4`) on the dataset's underlying array. NumPy raises `UFuncTypeError` for in-place true division when the target array has an integer dtype, so any Palmer run with integer-valued precipitation, PET, or AWC input (a plausible real-world netCDF encoding) would crash instead of computing.
- **Applied fix:** Switched to out-of-place division (`dataset[var].values = dataset[var].values / 25.4`) for precipitation, PET, and AWC, which promotes integer input to float instead of raising. Commit `1f99758`.

#### [P3] Parametrized Palmer test dropped coverage of the documented absent-`units` AWC path — Fixed
- **Location:** `tests/test_cli_e2e.py` (`test_palmers_writes_all_four_outputs_matching_in_process_computation`)
- **Origin:** Test-quality lane
- **Problem:** The CLI treats an AWC variable with no `units` attribute as already being in inches (a documented, supported input shape), but the parametrized test only exercised `awc_units` in `{"mm", "inches"}`, so a regression in that no-attribute branch would not be caught.
- **Applied fix:** Added `None` to the `awc_units` parametrization, writing an AWC variable with no `units` attribute in that case. Commit `1f99758`.

#### [P3] AWC-units rejection test didn't assert no output was written — Fixed
- **Location:** `tests/test_cli_e2e.py` (`test_palmers_rejects_an_awc_variable_with_unsupported_units`)
- **Origin:** Test-quality lane
- **Problem:** The file's other rejection test (`test_invalid_scale_raises_and_writes_no_output`) asserts both the raised exception and that no output file was written; the new AWC-rejection test only asserted the exception, so a regression that writes partial Palmer output before validation raises would go undetected.
- **Applied fix:** Added `assert not list(tmp_path.glob("palmers_*"))` after the `pytest.raises` block. Commit `cc590e52`.

#### [P4] AWC's accepted unit aliases were never exercised by a test — Fixed
- **Location:** `tests/test_cli_e2e.py` (`awc_units` parametrization, `_length_in` helper)
- **Origin:** Test-quality lane
- **Problem:** The AWC conversion accepts `mm`, `millimeters`, and `millimeter` as equivalent unit labels, but every test only ever wrote the literal string `"mm"`, leaving the other accepted aliases untested.
- **Applied fix:** Added `"millimeters"` to the `awc_units` parametrization and generalized `_length_in` accordingly. Commit `cc590e52`.

#### [P4] Whitespace-stripping is inconsistent between AWC and precip/PET/temperature unit matching — Not fixed (out of scope)
- **Location:** `src/climate_indices/__main__.py` (AWC unit check vs. the pre-existing precipitation/PET/temperature unit checks earlier in the same function)
- **Origin:** Correctness/edge-case lane
- **Problem:** The new AWC unit check strips whitespace before matching (e.g. `" mm "` is accepted); the pre-existing precipitation/PET/temperature checks do not, so the same whitespace would be rejected there with a whitespace-polluted error message.
- **Status:** Not fixed — the inconsistency is in code this PR does not touch (the precipitation/PET/temperature unit checks predate it and are shared by every other index, not just Palmers). Fixing it would broaden this PR beyond the Palmer units bug it targets. Flagging as a follow-up if unit-matching robustness is revisited generally.

### Remaining blockers
None.

File: src/climate_indices/__main__.py [hunk 1]
```diff
@@ -883,6 +883,12 @@ def _compute_write_index(keyword_arguments: dict[str, Any]) -> tuple[str, str] |
             else:
                 raise ValueError(f"Unsupported PET units: {dataset[pet_var_name].units}")
 
+    # the Palmer routines take inches, whereas the conversions above normalize
+    # precipitation and PET to millimeters for every other index
+    if keyword_arguments["index"] == "palmers":
+        dataset[keyword_arguments["var_name_precip"]].values /= 25.4
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/__main__.py]: Fixed in 1f997581 (a follow-up commit pushed shortly after this review, already on the branch): the Palmer conversions are now out-of-place (`dataset[var_name].values = dataset[var_name].values / 25.4`), so integer-backed precipitation/PET/AWC variables are promoted to float instead of raising a casting error.

[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/__main__.py]: Fixed in 15e4fc79: `--index palmers` now rejects a `mm/dy` precipitation label outright (`ValueError: Unsupported precipitation units for palmers: 'mm/dy' is a daily rate, not a monthly total`) instead of dividing it by 25.4 and treating it as a monthly depth. Covered by a new e2e test, `test_palmers_rejects_a_precipitation_rate_label`.

File: src/climate_indices/__main__.py [hunk 2]
```diff
@@ -909,9 +915,17 @@ def _compute_write_index(keyword_arguments: dict[str, Any]) -> tuple[str, str] |
...
         awc_dataset = xr.open_dataset(keyword_arguments["netcdf_awc"])
 
+        # the Palmer routines take available water capacity in inches; an
+        # absent units attribute is assumed to already be inches
+        var_name = keyword_arguments["var_name_awc"]
+        awc_units = str(awc_dataset[var_name].attrs.get("units", "")).strip().lower()
+        if awc_units in ("mm", "millimeters", "millimeter"):
+            awc_dataset[var_name].values /= 25.4
+        elif awc_units and awc_units not in ("inch", "inches"):
+            raise ValueError(f"Unsupported available water capacity units: {awc_units}")
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/__main__.py]: Fixed in 15e4fc79: the unsupported-units error now interpolates the AWC `units` attribute with `!r` (`f"Unsupported available water capacity units: {awc_units!r}"`), so embedded newlines/control characters are rendered as escaped text rather than raw.

File: src/climate_indices/__main__.py [hunk 3]
```diff
@@ -909,9 +915,17 @@ def _compute_write_index(keyword_arguments: dict[str, Any]) -> tuple[str, str] |
 
         awc_dataset = xr.open_dataset(keyword_arguments["netcdf_awc"])
 
+        # the Palmer routines take available water capacity in inches; an
+        # absent units attribute is assumed to already be inches
+        var_name = keyword_arguments["var_name_awc"]
+        awc_units = str(awc_dataset[var_name].attrs.get("units", "")).strip().lower()
+        if awc_units in ("mm", "millimeters", "millimeter"):
+            awc_dataset[var_name].values /= 25.4
+        elif awc_units and awc_units not in ("inch", "inches"):
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/__main__.py]: Fixed in 15e4fc79: the AWC file open + units validation (and the AWC inches conversion) now runs in the same palmers branch as the precipitation/PET conversion, before `_drop_data_into_shared_arrays_*` copies precipitation and PET into shared memory. An unsupported AWC label is now rejected before those full-array copies happen.
